File: problems/3508_implement_router/3508_implement_router.java
class Router {

    private final int memoryLimit;
    private final Deque<Packet> queue;
    private final Set<Packet> seen;
    private final Map<Integer, List<Integer>> destTimestamps;
    private final Map<Integer, Integer> processedIndex;


    public Router(int memoryLimit) {
                this.memoryLimit = memoryLimit;
        this.queue = new ArrayDeque<>();
        this.seen = new HashSet<>();
        this.destTimestamps = new HashMap<>();
        this.processedIndex = new HashMap<>();
    }
    
    public boolean addPacket(int source, int destination, int timestamp) {
        Packet p = new Packet(source, destination, timestamp);
        if (seen.contains(p)) return false;
        if (queue.size() == memoryLimit) {
            forwardPacket();
        }
        queue.addLast(p);
        seen.add(p);
        destTimestamps.computeIfAbsent(destination, k -> new ArrayList<>()).add(timestamp);
        return true;
    }
    
    public int[] forwardPacket() {
        if (queue.isEmpty()) return new int[]{};
        Packet p = queue.removeFirst();
        seen.remove(p);
        processedIndex.put(p.destination, processedIndex.getOrDefault(p.destination, 0) + 1);
        return new int[]{p.source, p.destination, p.timestamp};
    }
    
    public int getCount(int destination, int startTime, int endTime) {
        List<Integer> lst = destTimestamps.get(destination);
        if (lst == null) return 0;
        int left = processedIndex.getOrDefault(destination, 0);
        int lo = lowerBound(lst, left, startTime);
        int hi = upperBound(lst, lo, endTime);
        return hi - lo;
    }

    private int lowerBound(List<Integer> a, int lo, int target) {
        int l = lo, r = a.size();
        while (l < r) {
            int m = (l + r) >>> 1;
            if (a.get(m) < target) l = m + 1;
            else r = m;
        }
        return l;
    }
    private int upperBound(List<Integer> a, int lo, int target) {
        int l = lo, r = a.size();
        while (l < r) {
            int m = (l + r) >>> 1;
            if (a.get(m) <= target) l = m + 1;
            else r = m;
        }
        return l;
    }

    private static class Packet {
        int source, destination, timestamp;
        Packet(int s, int d, int t) { source = s; destination = d; timestamp = t; }

        @Override
        public boolean equals(Object o){
            if (this == o) return true;
            if (!(o instanceof Packet)) return false;
            Packet p = (Packet)o;
            return source == p.source && destination == p.destination && timestamp == p.timestamp;
        }
        @Override
        public int hashCode(){
            return Objects.hash(source, destination, timestamp);
        }
    }
}

/**
 * Your Router object will be instantiated and called as such:
 * Router obj = new Router(memoryLimit);
 * boolean param_1 = obj.addPacket(source,destination,timestamp);
 * int[] param_2 = obj.forwardPacket();
 * int param_3 = obj.getCount(destination,startTime,endTime);
 */

File: problems/3508_implement_router/3508_implement_router.py
from collections import deque, defaultdict
import bisect

class Router:
    def __init__(self, memoryLimit: int):
        self.memory_limit = memoryLimit
        self.queue = deque()  # stores tuples (source, destination, timestamp)
        self.seen = set()     # set of tuples to avoid duplicates
        self.dest_timestamps = defaultdict(list)   # destination -> list of timestamps (sorted by append)
        self.processed_index = defaultdict(int)    # destination -> number of timestamps popped/forwarded

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:
        key = (source, destination, timestamp)
        if key in self.seen:
            return False
        if len(self.queue) == self.memory_limit:
            self.forwardPacket()  # free one slot (problem requires this)
        self.queue.append(key)
        self.seen.add(key)
        self.dest_timestamps[destination].append(timestamp)
        return True

    def forwardPacket(self) -> list:
        if not self.queue:
            return []
        source, destination, timestamp = self.queue.popleft()
        self.seen.remove((source, destination, timestamp))
        # mark that destination lost one earliest timestamp
        self.processed_index[destination] += 1
        return [source, destination, timestamp]

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:
        lst = self.dest_timestamps.get(destination)
        if not lst:
            return 0
        left = self.processed_index.get(destination, 0)
        # use bisect with lo=left to ignore already forwarded timestamps
        lo = bisect.bisect_left(lst, startTime, lo=left)
        hi = bisect.bisect_right(lst, endTime, lo=left)
        return hi - lo
